File: mobile/src/main/java/com/siliconlabs/bledemo/ble/values/TemperatureValue.java
package com.siliconlabs.bledemo.ble.values;

import android.bluetooth.BluetoothGattCharacteristic;

/**
 * Gatt Characteristic value representing a temperature measurement.
 */
public class TemperatureValue {
    public final float temperature;
    public final boolean isFahrenheit;

    public final Integer temperatureType;
    public final int[] timeStamp;

    public static class Factory implements ValueFactory<TemperatureValue> {
        @Override
        public TemperatureValue create(BluetoothGattCharacteristic value) {
            return new TemperatureValue(value);
        }
    }

    private TemperatureValue(BluetoothGattCharacteristic value) {
        // https://developer.bluetooth.org/gatt/characteristics/Pages/CharacteristicViewer.aspx?u=org.bluetooth.characteristic.temperature_measurement.xml
        final int flags = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0) & 0x00000007;
        temperature = value.getFloatValue(BluetoothGattCharacteristic.FORMAT_FLOAT, 1);
        isFahrenheit = (flags & 0x00000001) != 0;

        switch (flags) {
            case 2:
            case 3:
                timeStamp = getTimeStamp(value, 1 + 4);
                temperatureType = null;
                break;

            case 4:
            case 5:
                timeStamp = null;
                // https://developer.bluetooth.org/gatt/characteristics/Pages/CharacteristicViewer.aspx?u=org.bluetooth.characteristic.temperature_type.xml
                temperatureType = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 1 + 4);
                break;

            case 6:
            case 7:
                timeStamp = getTimeStamp(value, 1 + 4);
                temperatureType = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 1 + 4 + 7);
                break;

            default:
                timeStamp = null;
                temperatureType = null;
                break;

        }
    }

    private int[] getTimeStamp(BluetoothGattCharacteristic value, int offset) {
        int year, month, day, hours, minutes, seconds;
        year = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT16, offset);
        month = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 2);
        day = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 3);
        hours = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 4);
        minutes = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 5);
        seconds = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 6);

        return new int[]{year, month, day, hours, minutes, seconds};
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Bluetooth/BLE/Values/TemperatureValue.kt
package com.siliconlabs.bledemo.bluetooth.ble.values

import android.bluetooth.BluetoothGattCharacteristic

/**
 * Gatt Characteristic value representing a temperature measurement.
 */
class TemperatureValue private constructor(value: BluetoothGattCharacteristic) {
    private var temperatureType: Int? = null
    private val timeStamp: IntArray?

    private val isFahrenheit: Boolean
    val temperature: Float

    class Factory : ValueFactory<TemperatureValue?> {
        override fun create(value: BluetoothGattCharacteristic): TemperatureValue {
            return TemperatureValue(value)
        }
    }

    init {
        val flags = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0) and 0x00000007
        temperature = value.getFloatValue(BluetoothGattCharacteristic.FORMAT_FLOAT, 1)
        isFahrenheit = flags and 0x00000001 != 0
        when (flags) {
            2, 3 -> {
                timeStamp = getTimeStamp(value, 1 + 4)
                temperatureType = null
            }
            4, 5 -> {
                timeStamp = null
                temperatureType = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 1 + 4)
            }
            6, 7 -> {
                timeStamp = getTimeStamp(value, 1 + 4)
                temperatureType = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 1 + 4 + 7)
            }
            else -> {
                timeStamp = null
                temperatureType = null
            }
        }
    }

    private fun getTimeStamp(value: BluetoothGattCharacteristic, offset: Int): IntArray {
        val year: Int = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT16, offset)
        val month: Int = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 2)
        val day: Int = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 3)
        val hours: Int = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 4)
        val minutes: Int = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 5)
        val seconds: Int = value.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, offset + 6)
        return intArrayOf(year, month, day, hours, minutes, seconds)
    }

}